calico_kubernetes/calico_kubernetes.py

```diff
@@ -181,37 +225,64 @@ def _container_add(self, pid, ip, interface, hostname, orchestrator_id):
-                         "pools. Exiting.", ip)
-            sys.exit(1)
-        else:
-            if not ip_assigned:
-               logger.error("Failed to assign IP address %s. Exiting.", ip)
-               sys.exit(1)
+        ip_list = self._assign_container_ip()
 
         # Create Endpoint object
         try:
+            logger.info("Creating endpoint for ips in list %s", ip_list)
             ep = self._datastore_client.create_endpoint(hostname, orchestrator_id,
-                                                        self.docker_id, [ip])
+                                                        self.docker_id, ip_list)
         except AddrFormatError:
             logger.error("This node is not configured for IPv%d. Unassigning "
-                         "IP address %s then exiting.", ip.version, ip)
-            self._datastore_client.unassign_address(None, ip)
+                         "IP address %s then exiting.", ip_list[0].version, ip_list[0])
+            self._datastore_client.release_ips(set(ip_list))
             sys.exit(1)
 
         # Create the veth, move into the container namespace, add the IP and
         # set up the default routes.
-        ns = netns.PidNamespace(pid)
-        ep.mac = ep.provision_veth(ns, interface)
+        logger.info("Creating the veth with namespace pid %s on interface name %s", pid, interface)
+        ep.mac = ep.provision_veth(netns.PidNamespace(pid), interface)
+
+        logger.info("Setting mac address %s to endpoint %s", ep.mac, ep.name)
         self._datastore_client.set_endpoint(ep)
 
         # Let the caller know what endpoint was created.
         return ep
 
+    def _assign_container_ip(self):
+        """
+        Assign IPAddress either with the assigned docker IPAddress or utilize
+        calico IPAM.
+
+        The option to utilize IPAM is indicated by the environment variable
+        "CALICO_IPAM".
+        True indicates to utilize Calico's auto_assign IPAM policy.
+        False indicate to utilize the docker assigned IPAddress
+
+        :return list containing the IPAddress which has been assigned
+        """
+        if CALICO_IPAM == 'true':
+            # Assign ip address through IPAM Client
+            logger.info("Using Calico IPAM")
+            try:
+                ip_list, ipv6_addrs = self._datastore_client.auto_assign_ips(
+                    1, 0, None, None)
+                logger.info("ip_list is %s; ipv6_addrs is %s", ip_list, ipv6_addrs)
+                assert not ipv6_addrs
+            except RuntimeError as err:
+                logger.error("Cannot auto assign IPAddress: %s", err.message)
+                sys.exit(1)
+        else:
+            logger.info("Using docker assigned IP address")
+            ip_list = [self._read_docker_ip()]
+            try:
+                self._datastore_client.assign_ip(ip_list[0], str(self.docker_id), None)
+            except Exception:
```

Comment:
Don't catch all exceptions here. In general, its bad to do "Pokemon style" exception handling (gotta catch 'em all), since it can hide bugs.  Just catch what you expect to be thrown.